docs/designs/CodeFlare-SDK-design-doc.md

```diff
@@ -0,0 +1,141 @@
+7. Follow best practices in codebase to make it easier for developers to maintain the project.
+
+## Architecture
+
+### CodeFlare stack components overview:
+1. [CodeFlare SDK][codeflare_sdk_gh]: Primary user-facing component, offering a Python-native interface that bridges users with the underlying services in the stack. This SDK abstracts away much of the complexity, providing a unified and accessible method for submitting and managing the running of distributed AI workloads on Kubernetes clusters.
+
+2. [CodeFlare Operator][codeflare_operator_gh]: Manages the platform requirements for the running of the CodeFlare components:
+   1. Ensuring necessary configurations are present on cluster to enable access to Ray dashboards (includes ingress and cert management).
+   2. Validate and mutate Ray cluster/AppWrapper CRUD requests (admission policy).
+   3. Creates Ray related Routes/Ingresses based on if the environment is OpenShift/Kubernetes.
+
+3. [AppWrapper][appwrapper_gh]: AppWrapper groups multiple Kubernetes resources into a single, logical workload for easier management. It is integrated with Kueue and accepts the k8s components to be created and managed through a template in the spec which is untyped in the form of RawExtension.
+
+4. [KubeRay][kuberay_gh]: Operator designed specifically for managing and orchestrating Ray clusters on Kubernetes. It automates the creation and lifecycle management of Ray clusters and jobs, ensuring that they can scale and operate efficiently on K8s. KubeRay also provides a Python client library, enabling developers and data scientists to interact with Ray components directly within their applications.
+
+5. [Kueue][kueue_gh]: Batch workload queuing and scheduling system designed to optimize resource allocation in K8s clusters. It ensures that Ray Clusters are only scheduled when sufficient resources are available, preventing resource contention and maximizing cluster efficiency. This component plays a critical role in balancing resource availability with the demand for distributed workloads, facilitating efficient job execution.
+
+On a high level, the interaction between a data scientist, CodeFlare SDK and the rest of the components are explained as below:
+
+![Diagram-1](../images/codeflare_stack_arch.png)
+
+The role of the SDK in the model training and tuning stack remains consistent on the client side, regardless of any changes to the components it interacts with on the cluster.
+
+![Diagram-2](../images/codeflare_sdk.png)
+
+#### Considerations while designing SDK:
+1. Ease of use and integration: The SDK’s primary role is to abstract Kubernetes specifics. It should provide simple interfaces for interacting with any of the model training components on the server side.
+2. Lightweight: The SDK runs client-side and should minimize resource consumption. It must prioritize responsiveness and user experience. For example, using a polling mechanism to fetch status instead of a constant resource watch.
```

Comment:
```suggestion
2. Lightweight: The SDK runs client-side and should minimize resource consumption. It must prioritize responsiveness and user experience. For example, using a polling mechanism to fetch status instead of constantly watching resources.
```